book/src/development/emitting_lints.md

```diff
@@ -0,0 +1,217 @@
+
+A suggestion is a change that the lint provides to fix the issue it is linting.
+The output looks something like this (from the example earlier):
+
+```text
+error: an inclusive range would be more readable
+  --> $DIR/range_plus_minus_one.rs:37:14
+   |
+LL |     for _ in 1..1 + 1 {}
+   |              ^^^^^^^^ help: use: `1..=1`
+```
+
+**Not all suggestions are always right**, some of them require human supervision, that's why we have
+[Applicability][applicability].
+
+Applicability indicates confidence in the correctness of the suggestion, some are always right
+(`Applicability::MachineApplicable`), but we use `Applicability::MaybeIncorrect` and other when talking about a lint
+that may be incorrect.
+
+---
+
+The same lint (`foo_functions`) but that emits a suggestion would be something like this:
+
+```rust
+impl<'tcx> LateLintPass<'tcx> for BarExpressions {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>)  {
+        // Imagine that `some_bar_expr_logic` checks for requirements for emitting the lint
+        if some_bar_expr_logic(expr) {
+            span_lint_and_sugg( // < Note this change
+                cx,
+                BAR_EXPRESSIONS,
+                span,
+                "message on why the lint is emitted",
+                "use",
+                "suggestion (don't forget to integrate things from the source, like variable names)", // < Suggestion
+                Applicability::MachineApplicable
+            );
+        }
+    }
+}
+```
+
+Suggestions generally use the [`format!`](format_macro) macro to interpolate the old values with the new ones.
+
+## How to choose between notes, help messages and suggestions
+
+Notes are presented separately from the main lint message, they provide useful information that the user needs to
+understand why the lint was activated. They are the most helpful when attached to a span.
+
+Example:
+
+```text
+error: calls to `std::mem::forget` with a reference instead of an owned value. Forgetting a reference does nothing.
+  --> $DIR/drop_forget_ref.rs:10:5
+   |
+10 |     forget(&SomeStruct);
+   |     ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: `-D clippy::forget-ref` implied by `-D warnings`
+note: argument has type &SomeStruct
+  --> $DIR/drop_forget_ref.rs:10:12
+   |
+10 |     forget(&SomeStruct);
+   |            ^^^^^^^^^^^
+```
+
+---
+
+Help messages are specifically to help the user. These are used in situation where you can't provide a specific
+machine applicable suggestion. They can also be attached to a span.
+
+Example:
+
+```text
+error: constant division of 0.0 with 0.0 will always result in NaN
+  --> $DIR/zero_div_zero.rs:6:25
+   |
+6  |     let other_f64_nan = 0.0f64 / 0.0;
+   |                         ^^^^^^^^^^^^
+   |
+   = help: consider using `f64::NAN` if you would like a constant representing NaN
+```
+
+---
+
+Suggestions are the most helpful, they are direct changes to the source code in order to fix the error. The magic
```

Comment:
They are not really direct changes, but rather code changes that can be applied with `rustfix`.